### PR TITLE
Implement the reconciliation loop for GCPManagedMachinePool (GKE Part 4)

### PR DESCRIPTION
**Pull request by richardchen331**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This change adds the reconciliation logic for the GCPManagedControlPlane.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Relates https://github.com/kubernetes-sigs/cluster-api-provider-gcp/issues/512

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Add the reconciliation logic for **GCPManagedMachinePool**
```

### CODE REVIEW

File: config/manager/manager.yaml [hunk 1]
```diff
@@ -21,6 +21,7 @@ spec:
       containers:
       - args:
         - --leader-elect
+        - --feature-gates=GKE=${EXP_CAPG_GKE:=false}
```
Comment (richardcase) [inline config/manager/manager.yaml:24]: Can we remove this until PR5?

File: exp/controllers/gcpmanagedmachinepool_controller.go [hunk 2]
```diff
@@ -29,33 +57,320 @@ import (
...
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the GCPManagedMachinePool object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.2/pkg/reconcile
-func (r *GCPManagedMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+func (r *GCPManagedMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
```
Comment (richardcase) [inline exp/controllers/gcpmanagedmachinepool_controller.go:225]: We'll need permissions for MachinePool, Cluster and GCPManagedCluster, GCPManagedControlPlane

File: feature/feature.go [hunk 3]
```diff
@@ -20,12 +20,16 @@ import (
 )
 
 const (
-// Every capg-specific feature gate should add method here following this template:
-//
-// // owner: @username
-// // alpha: v1.X
-// MyFeature featuregate.Feature = "MyFeature".
+	// Every capg-specific feature gate should add method here following this template:
```
Comment (richardcase) [inline feature/feature.go:23]: We can remove this as it will be added in PR5

Comment (richardcase) [reply feature/feature.go:23]: I'm also wondering do we need to worry about the CAPI machinepool feature flag?

Comment (richardcase) [reply feature/feature.go:23]: But we can worry about that in the next PR :)

File: main.go [hunk 4]
```diff
@@ -143,56 +147,117 @@ func main() {
...
 	}).SetupWithManager(ctx, mgr, controller.Options{MaxConcurrentReconciles: gcpClusterConcurrency}); err != nil {
-		setupLog.Error(err, "unable to create controller", "controller", "GCPCluster")
-		os.Exit(1)
+		return fmt.Errorf("setting up GCPCluster controller: %w", err)
 	}
 
-	if err = (&infrav1beta1.GCPCluster{}).SetupWebhookWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create webhook", "webhook", "GCPCluster")
-		os.Exit(1)
+	if feature.Gates.Enabled(feature.GKE) {
```
Comment (richardcase) [inline main.go:189]: We'll do this in the next PR when we bring it all together.

File: main.go [hunk 5]
```diff
@@ -143,56 +147,117 @@ func main() {
...
+	}
+	if err := (&infrav1beta1.GCPMachine{}).SetupWebhookWithManager(mgr); err != nil {
+		return fmt.Errorf("setting up GCPMachine webhook: %w", err)
+	}
+	if err := (&infrav1beta1.GCPMachineTemplate{}).SetupWebhookWithManager(mgr); err != nil {
+		return fmt.Errorf("setting up GCPMachineTemplate webhook: %w", err)
+	}
+
+	if feature.Gates.Enabled(feature.GKE) {
+		if err := (&infrav1exp.GCPManagedCluster{}).SetupWebhookWithManager(mgr); err != nil {
```
Comment (richardcase) [inline main.go:233]: Same comment as above with enabling the GKE controllers in the next PR

**Review by richardcase — COMMENTED** [review]
Thanks @richardchen331 🎉

File: cloud/scope/managedmachinepool.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+// ParseInstanceGroupURL parses an instance group URL.
+func ParseInstanceGroupURL(url string) (project string, zone string, mig string) {
+	parts := strings.Split(url, "/")
+	if len(parts) < 11 {
+		return "", "", ""
+	}
+	return parts[6], parts[8], parts[10]
+}
+
+// ConvertToSdkNodePool convertsSetReplicas a node pool to format that is used by GCP SDK.
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/scope/managedmachinepool.go:165]: fixed

File: cloud/scope/managedmachinepool.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+		Name:             nodePoolName,
+		InitialNodeCount: nodePool.Spec.NodeCount,
+		Config: &containerpb.NodeConfig{
+			Labels:   nodePool.Spec.KubernetesLabels,
+			Taints:   infrav1exp.ConvertToSdkTaint(nodePool.Spec.KubernetesTaints),
+			Metadata: nodePool.Spec.AdditionalLabels,
+		},
+	}
+	if nodePool.Spec.NodeVersion != nil {
+		sdkNodePool.Version = *nodePool.Spec.NodeVersion
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/scope/managedmachinepool.go:181]: make sense, updated

File: exp/controllers/gcpmanagedmachinepool_controller.go (hunk 2 above)
Comment (richardchen331) [reply exp/controllers/gcpmanagedmachinepool_controller.go:225]: updated

File: config/manager/manager.yaml (hunk 1 above)
Comment (richardchen331) [reply config/manager/manager.yaml:24]: removed

File: feature/feature.go (hunk 3 above)
Comment (richardchen331) [reply feature/feature.go:23]: change reverted :)

File: main.go (hunk 4 above)
Comment (richardchen331) [reply main.go:189]: reverted

File: main.go (hunk 5 above)
Comment (richardchen331) [reply main.go:233]: reverted

File: cloud/scope/managedmachinepool.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+// InstanceGroupManagersClient returns a client used to interact with GCP MIG.
+func (s *ManagedMachinePoolScope) InstanceGroupManagersClient() *compute.InstanceGroupManagersClient {
+	return s.migClient
+}
+
+// ParseInstanceGroupURL parses an instance group URL.
+func ParseInstanceGroupURL(url string) (project string, zone string, mig string) {
+	parts := strings.Split(url, "/")
+	if len(parts) < 11 {
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/scope/managedmachinepool.go:159]: updated

File: cloud/scope/managedcontrolplane.go [hunk 9]
```diff
@@ -171,18 +173,33 @@ func (s *ManagedControlPlaneScope) GetCredential() *Credential {
...
+			return nil, err
+		}
+		s.AllNodePools = machinePoolList.Items
 	}
-	return region, nil
+
+	return s.AllNodePools, nil
+}
+
+func parseLocation(location string) string {
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/scope/managedcontrolplane.go:206]: updated

**Review by richardcase — COMMENTED** [review]
Fantastic @richardchen331

File: cloud/scope/managedcontrolplane.go [hunk 10]
```diff
@@ -171,19 +176,36 @@ func (s *ManagedControlPlaneScope) GetCredential() *Credential {
...
 
-func parseLocation(location string) (region string, zone *string) {
-	parts := strings.Split(location, "-")
-	region = strings.Join(parts[:2], "-")
-	if len(parts) == 3 {
-		return region, &parts[2]
+// GetAllNodePools gets all node pools for the control plane.
+func (s *ManagedControlPlaneScope) GetAllNodePools(ctx context.Context) ([]infrav1exp.GCPManagedMachinePool, []clusterv1exp.MachinePool, error) {
+	if s.AllManagedMachinePools == nil || len(s.AllManagedMachinePools) == 0 {
+		opt1 := client.InNamespace(s.GCPManagedControlPlane.Namespace)
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/scope/managedcontrolplane.go:182]: updated

File: cloud/services/container/clusters/kubeconfig.go [hunk 11]
```diff
@@ -198,7 +198,7 @@ func (s *Service) updateCAPIKubeconfigSecret(ctx context.Context, configSecret *
 }
 
 func (s *Service) getKubeConfigContextName(isUser bool) string {
-	contextName := fmt.Sprintf("gke_%s_%s_%s", s.scope.GCPManagedControlPlane.Spec.Project, s.scope.GCPManagedControlPlane.Spec.Location, s.scope.GCPManagedControlPlane.Name)
+	contextName := fmt.Sprintf("gke_%s_%s_%s", s.scope.GCPManagedControlPlane.Spec.Project, s.scope.GCPManagedControlPlane.Spec.Location, s.scope.GCPManagedControlPlane.Spec.ClusterName)
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/services/container/clusters/kubeconfig.go:201]: sg! updated

File: cloud/services/container/clusters/reconcile.go [hunk 12]
```diff
@@ -190,18 +215,14 @@ func (s *Service) describeCluster(ctx context.Context) (*containerpb.Cluster, er
 func (s *Service) createCluster(ctx context.Context) error {
 	log := log.FromContext(ctx)
 
+	nodePools, machinePools, _ := s.scope.GetAllNodePools(ctx)
 	cluster := &containerpb.Cluster{
-		Name:    s.scope.GCPManagedControlPlane.Name,
+		Name:    s.scope.GCPManagedControlPlane.Spec.ClusterName,
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/services/container/clusters/reconcile.go:220]: agreed, updated

File: cloud/services/container/nodepools/reconcile.go [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+	log.Info("Reconciling node pool resources")
+
+	nodePool, err := s.describeNodePool(ctx)
+	if err != nil {
+		s.scope.GCPManagedMachinePool.Status.Ready = false
+		conditions.MarkFalse(s.scope.ConditionSetter(), clusterv1.ReadyCondition, infrav1exp.GKEMachinePoolReconciliationFailedReason, clusterv1.ConditionSeverityError, err.Error())
+		return ctrl.Result{}, err
+	}
+	if nodePool == nil {
+		log.Info("Node pool not found, creating")
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/services/container/nodepools/reconcile.go:54]: done

File: cloud/services/container/nodepools/reconcile.go [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+	if needUpdateVersionOrImage {
+		log.Info("Version/image update required")
+		err = s.updateNodePoolVersionOrImage(ctx, nodePoolUpdateVersionOrImage)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		log.Info("Node pool version/image updating in progress")
+		s.scope.GCPManagedMachinePool.Status.Ready = true
+		conditions.MarkTrue(s.scope.ConditionSetter(), infrav1exp.GKEMachinePoolUpdatingCondition)
+		return ctrl.Result{}, nil
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/services/container/nodepools/reconcile.go:127]: you're right, we should requeue here

File: cloud/services/container/nodepools/reconcile.go [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+	if needUpdateSize {
+		log.Info("Size update required")
+		err = s.updateNodePoolSize(ctx, setNodePoolSizeRequest)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		log.Info("Node pool size updating in progress")
+		s.scope.GCPManagedMachinePool.Status.Ready = true
+		conditions.MarkTrue(s.scope.ConditionSetter(), infrav1exp.GKEMachinePoolUpdatingCondition)
+		return ctrl.Result{}, nil
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/services/container/nodepools/reconcile.go:140]: make sense, updated

File: cloud/services/container/nodepools/reconcile.go [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+		return ctrl.Result{}, err
+	}
+	if nodePool == nil {
+		log.Info("Node pool already deleted")
+		conditions.MarkFalse(s.scope.ConditionSetter(), infrav1exp.GKEMachinePoolDeletingCondition, infrav1exp.GKEMachinePoolDeletedReason, clusterv1.ConditionSeverityInfo, "")
+		return ctrl.Result{}, err
+	}
+
+	switch nodePool.Status {
+	case containerpb.NodePool_PROVISIONING:
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/services/container/nodepools/reconcile.go:171]: i think GKE doesn't allow you to perform any operation (e.g. delete) while one is still in progress (e.g. provisioning), but we should requeue here.

File: cloud/services/container/nodepools/reconcile.go [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+		log.Info("Node pool provisioning in progress")
+		return ctrl.Result{}, nil
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply cloud/services/container/nodepools/reconcile.go:173]: updated

File: exp/api/v1beta1/gcpmanagedmachinepool_types.go [hunk 18]
```diff
@@ -30,13 +30,13 @@ const (
...
+	// NodePoolName specifies the name of the GKE node pool corresponding to this MachinePool. If you don't specify a name
+	// then a default name will be created based on the namespace and name of the managed machine pool.
 	// +optional
-	NodeVersion *string `json:"nodeVersion,omitempty"`
-	// InitialNodeCount represents the initial number of nodes for the pool.
+	NodePoolName string `json:"nodePoolName,omitempty"`
+	// NodeCount represents the initial number of nodes for the pool.
 	// In regional or multi-zonal clusters, this is the number of nodes per zone.
-	InitialNodeCount int32 `json:"initialNodeCount"`
+	NodeCount int32 `json:"nodeCount"`
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply exp/api/v1beta1/gcpmanagedmachinepool_types.go:39]: my plan is to get an MVP working and adding more features after the initial set of PRs are merged. let me know if that make sense :)

[the remark this answers is not in the document]
Comment (richardcase) [reply exp/api/v1beta1/gcpmanagedmachinepool_types.go:39]: I would agree with this approach for most features. But I wonder as this is a machinepool should we have scaling from the beginning? Feels like we should but happy to go with what you feel is best,

**Comment by richardcase** [comment]
@richardchen331 - sorry for the late replay. I left 1 reply around auto-scaling. Apart from that, I think this looks good. Could you rebase and squash commits?

**Comment by richardchen331** [comment]
> @richardchen331 - sorry for the late replay. I left 1 reply around auto-scaling. Apart from that, I think this looks good. Could you rebase and squash commits?

Thanks for the review! I rebased and squashed the commits (only left one additional for the autoscaling change), and add auto scaling support. Please take another look when you have time. I can squash again after it's approved :)

**Comment by richardcase** [comment]
@richardchen331 - this looks great, and thanks for adding the autoscaling options initially. I think this will help. From my side, i have no further comments on the code.

Can you squash and handle the conflicting `go.mod`? 

I will then approve :)

**Comment by richardchen331** [comment]
> @richardchen331 - this looks great, and thanks for adding the autoscaling options initially. I think this will help. From my side, i have no further comments on the code.
> 
> Can you squash and handle the conflicting `go.mod`?
> 
> I will then approve :)

Sure. I rebased and squashed the commits. Could you help approve the PR? Thanks Richard!

**Review by cpanato — APPROVED** [review]
nice work

/approve

**Pull request merged**